clap_complete/src/dynamic/completer.rs

```diff
@@ -64,18 +65,82 @@ pub fn complete(
             if let Some(next_cmd) = current_cmd.find_subcommand(value) {
                 current_cmd = next_cmd;
                 pos_index = 1;
+                _state = ParseState::ValueDone;
                 continue;
             }
         }
 
         if is_escaped {
             pos_index += 1;
+            _state = ParseState::Pos(pos_index);
         } else if arg.is_escape() {
             is_escaped = true;
-        } else if let Some(_long) = arg.to_long() {
-        } else if let Some(_short) = arg.to_short() {
+            _state = ParseState::ValueDone;
+        } else if let Some((flag, value)) = arg.to_long() {
+            if let Ok(flag) = flag {
+                _state = if let Some(opt) = current_cmd.get_arguments().find(|a| {
+                    a.get_long_and_visible_aliases()
+                        .map_or(false, |v| v.into_iter().find(|s| *s == flag).is_some())
+                }) {
+                    match opt.get_action() {
+                        clap::ArgAction::Set | clap::ArgAction::Append => {
+                            if value.is_some() {
+                                ParseState::ValueDone
+                            } else {
+                                ParseState::Opt(opt.clone())
+                            }
+                        }
+                        _ => {
+                            if value.is_some() {
+                                ParseState::Unknown
+                            } else {
+                                ParseState::ValueDone
+                            }
+                        }
+                    }
+                } else {
+                    ParseState::Unknown
+                };
+            }
+        } else if let Some(short) = arg.to_short() {
+            let mut short = short.clone();
+            let opt = short.next_flag();
+
```

Comment:
I thought this PR was supposed to just  be a refactor that added `ParseState::ValueDone` and `ParseState::Pos`.  I believe the intention was to remove the use of passing `is_escaped` to `complete_arg`.

Reply:
This PR is supposed to add a struct to record the context during parsing. Actually, `ParseState::Opt` is the key filed in this PR. We could use it to complete `--flag bar` and `-f bar`. And i think that `ParseState` could be expanded if there are other contexts.
>I believe the intention was to remove the use of passing is_escaped to complete_arg.

Yes, we can do this. `is_escaped` is like a context variable for the word after `--`

Reply:
>The PR / commit title say this is a feature but this looks like its a refactor (code clean up with no user-visible change). Could you clarify what the end-user impact of this either in the type or in the description?

This PR does not provide new features for users, but it provides a feature for developers to record more context. From the user's perspective, this PR should perhaps be labeled as a refactor.

Reply:
> This PR is supposed to add a struct to record the context during parsing.

That is not the reason I suggested making this PR and this not a mergeable direction for this PR.  Commits need to be atomic which this is not.  PRs can either be a single commit or multiple that "tell a story together".

> a feature for developers to record more context. 

That is a refactor, not a feature.  The terms are user-facing.

Reply:
I don't think this counts as a refactor either. it's the codebase for the new feature https://github.com/clap-rs/clap/pull/5539. The reason I separated it is that I think this PR could may also serve as the code base for other features, if someone wants to support multivalue completion, he could add a field in `ParseState` and add some logic in `complete` function.
From the user's perspective, this PR is indeed not atomic. So perhaps I should merge the code from PR https://github.com/clap-rs/clap/pull/5539 into this PR and then close it? I'll keep the two commits to make it easier for everyone to track.